Add version sort for variables

hasso opened the pull request:
Adds support for "version" sorting for variables. It's not bulletproof and doesn't work with very long numbers, but it's simple, fast and works with all kind of separators. It should be enough for all common use cases – network interface names, IPv4-addresses encoded into hostnames, version numbers of course etc.

Implements the feature request #7307.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/grafana/grafana?pullRequest=11628) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/grafana/grafana?pullRequest=11628) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/grafana/grafana?pullRequest=11628) it.</sub>

dehrax commented:
Thanks for the PR! :)

It feels like this functionality would be perfect to have some test cases for, so we know a little bit better what "bulleproof" includes. Could you fix that? :)

hasso committed:
Make numeric sort much more powerful

Adds support for sorting based on numbers in arbitrary places in
variables and with variable number of digits. The pattern most
commonly used as version numbers - e.g 'pre-0.0.1, pre-0.0.2,
pre-0.0.18', 'pre-0.0.21' - but it's also common in network
addresses, interface names etc.

Th idea is simple - every number with up to 15 digits in the source
string is replaced with a number with 16 digits (10^15 + number)
and the result is used as a sorting key. 16 is choosen because it's
a maximum number of digits which is safe to do math with
(MAX_SAFE_INTEGER).

Every numeric case which was handled by previous code, is handled the
same way by the new one.

hasso commented:
I rewrote the code although the idea is same. The current numeric sort is very confusing from UI point of view anyway, therefore I thought that the new code should just replace it.

It breaks only one case illustrated by failing test as well. The new code doesn't ignore other symbols before the first number:

Old numeric sort: 'bbb2', 'ccc3', 'aaa10'
New numeric sort: 'aaa10', 'bbb2', 'ccc3'

Regarding bulletproofness – it can't think of any scenario where it'd break, but it MAY give you unexpected results if your strings contain numbers with more than 15 digits. For example this:

aaa00000000000000022
aaa21

Although 21 is smaller than 22, it sorted this way because of 15 zeros before 22. But I'd consider this as a real corner case.

Let me know if it's OK and I'll fix tests as well (and add some more). If it's not, I'll add a new sorting method back and will add tests for it. But it's really not worth of effort IMHO.
